proposals/tf-operator-design-v1alpha2.md

```diff
@@ -0,0 +1,435 @@
+)
+```
+
+**TFJobPhase**:
+```go
+// TFJobPhase is high-level summary of where the TFJob is in its lifecycle
+type TFJobPhase string
+
+const (
+    // The TFJob has been accepted by the Kubernetes system and TFJob controller,
+    // but one or more of the services/pods has not been created.
+    TFJobPending TFJobPhase = "Pending"
+
+    // All sub-resources (e.g. services/pods) of the TFJob has been successfully created,
+    // and have been successfully scheduled and launched.
+    // The training is running without error.
+    TFJobRunning TFJobPhase = "Running"
+
+    // All pods in the TFJob have terminated in success,
+    // and will not be restarted.
+    // The training is complete without error.
+    TFJobSucceeded TFJobPhase = "Succeeded"
+
+    // All pods in the TFJob have terminated, and at least one pod
+    // has terminated in failure.
+    // The training is failed.
+    TFJobFailed TFJobPhase = "Failed"
+
+    // For some reason the state of the TFJob could not be obtained,
+    // typically due to an error in communicating with the host of the TFJob.
+    TFJobUnknown TFJobPhase = "Unknown"
+)
+```
+
+## Design
+
+### TFController
+
+The TFJob controller `tf-operator` will process TFJobs and CRUD services/pods according to the spec of TFJob. It is responsible for synchronizing TFJob objects stored in the system with actual running services and pods, continuously strive to make the observed state match the desired state.
+
+Here is the definition of `TFController`:
+
+```go
+type TFController struct {
+    // kubeClientset is a standard kubernetes clientset
+    kubeClientset kubernetes.Interface
+
+    // tfJobClientset is a clientset for CRD TFJob
+    tfJobClientset tfjobclient.Interface
+
+    tfJobLister listers.TFJobLister
+    tfJobSynced cache.InformerSynced
+
+    // for pod/service CRUD
+    podLister     kubelisters.PodLister
+    podControl    controller.PodControlInterface
+    serviceLister kubelisters.ServiceLister
+
+    // workQueue is a rate limited work queue. This is used to queue work to be
+    // processed instead of performing it as soon as a change happens. This
+    // means we can ensure we only process a fixed amount of resources at a
+    // time, and makes it easy to ensure we are never processing the same item
+    // simultaneously in two different workers.
+    workQueue workqueue.RateLimitingInterface
+
+    // recorder is an event recorder for recording Event resources to the
+    // Kubernetes API.
+    recorder record.EventRecorder
+
+    // A TTLCache of pod creates/deletes each TFReplica expects to see
+    expectations controller.ControllerExpectationsInterface
+}
+```
+
+### Distributed TensorFlow Configuration
+
+**Auto-Generated Arguments:**
+
+To make distributed TensorFlow work, user **MUST** implement a parser in their code like `ArgumentParser` to get the distributed TensorFlow configurations which generated by `tf-operator`, the common built-in arguments are:
```

Comment:
I wonder if it's better to pass these in as environment variables instead of command line arguments. Seems it would be more flexible, as the user code can ignore these if they don't care about them, rather than being forced to parse them.

Reply:
I'd say we could do both, or even add config file into the mix - CLI overriding ENV overriding config file

Reply:
+1 to Rong's suggestion of using environment variables and more specifically the environment variable TF_CONFIG.

TensorFlow has already adopted the environment variabble TF_CONFIG
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/estimators/run_config.py#L84

as a convention for passing runtime information (e.g. # of workers) to libraries like tf.Estimator that can process that information.

Why would we try to define a new convention in the form of a set of specific command line arguments?

Passing command line arguments also seems brittle. What if existing code uses different flag parsing conventions e.g. "-" as opposed to "--"? What if code is using tf.Estimator and therefore can automatically use TF_CONFIG and strict argument parsing. Then the argument parser would raise an exception because of unrecognized command line arguments. So users would have to define arguments specific to TfJob operator even though the user's code was never using these arguments.
new convention in the form of a set of specific command line arguments?

Passing command line arguments also seems brittle. What if existing code uses different flag parsing conventions e.g. "-" as opposed to "--"?

Reply:
We have some discussions in https://github.com/kubeflow/tf-operator/issues/369:

> TF_CONFIG is a TensorFlow convention that TF APIs like the EstimatorAPI use to get information about the runtime environment and configure the job appropriately.
>
> If users want to use command line arguments, they can write a launcher script that parses TF_CONFIG and sets the command line arguments as needed. This is what we do in the TFCNN example; see https://github.com/kubeflow/kubeflow/blob/master/tf-controller-examples/tf-cnn/launcher.py
>
> I think the problem with command line arguments is that everyone will use slightly different conventions.

Reply:
@jlewi Hi, any tensorflow official docs about `TF_CONFIG`, I check the example from tensorflow website: https://www.tensorflow.org/deploy/distributed.

And i notice that `Run Config (deprecated, use tf.estimator.RunConfig instead).` in here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/estimators/run_config.py#L15

Reply:
And what it user don not use `estimator` library?

Reply:
https://www.tensorflow.org/api_docs/python/tf/estimator/RunConfig

The RunConfig docs say that it will automatically be populated based on the environment variable TF_CONFIG if it is set.

My hope is that most high level libraries will follow the convention of using the environment variable TF_CONFIG to automatically configure themselves.

As @gaocegege points out above. If a user is using a library that doesn't use TF_CONFIG then they have two options

1. Modify the code to read and parse TF_CONFIG
1. Create a launcher script (possibly using an init container) to parse TF_CONFIG and set command line arguments appropriately.

I really think we should try to avoid introducing new conventions (e.g. standard command line flags) when the TF community already has an existing convention.

Furthermore, when I look at TF code I don't see broad agreement about what command line flags should be used or how they should be structured.

Reply:
@jlewi Thanks, will take a deeper look at this later.

Reply:
@jlewi After investigating, `TF_CONFIG` is used for `tf.estimators`, which is a high level encapsulation. How about pass the both `TF_CONFIG` and `ClusterSpec` together?

- Generate `TF_CONFIG` JSON string and set it in `Command` field.
- Generate `ClusterSpec` args and set it in `Args` field.

So both estimators model and other tensorflow model can work together.

Reply:
@jlewi For `TF_CONFIG` i mean `Env` filed, sorry for the typo.